src/Data.Json.JSemantic.purs

```diff
@@ -0,0 +1,105 @@
+module Data.Json.JSemantic
+  ( JSemantic(..)
+  , toSemantic
+  ) where
+
+import Data.Argonaut.JCursor(JsonPrim(..), runJsonPrim)
+import Data.Maybe
+import Data.String.Regex (RegexFlags(..), regex, test)
+import Data.String
+import Data.Foldable
+import qualified Data.Array as A
+import qualified Data.Array.Unsafe as AU
+import Data.Tuple
+import qualified Data.Date as Date
+import Math
+
+
+data JSemantic = Integral   Number
+               | Fractional Number
+               | Date       Date.Date
+               | DateTime   Date.Date
+               | Time       Date.Date
+               | Interval   Date.Date Date.Date
+               | Text       String
+               | Bool       Boolean
+               | Percent    Number
+               | Currency   Number
+               | NA
+
+instance showJSemantic :: Show JSemantic where
+  show (Integral   n) = show n
+  show (Fractional n) = show n
+  show (Date       d) = show d
+  show (DateTime   d) = show d
+  show (Time       d) = show d
+  show (Interval u v) = show u ++ " - " ++ show v
+  show (Text       s) = s
+  show (Bool       b) = show b
+  show (Percent    n) = show n ++ "%"
+  show (Currency   n) = "$" ++ show n
+  show (NA)           = ""
+
+instance eqJSemantic :: Eq JSemantic where
+  (==) (Integral   a) (Integral   b) = a == b
+  (==) (Fractional a) (Fractional b) = a == b
+  (==) (Date       a) (Date       b) = a == b
+  (==) (DateTime   a) (DateTime   b) = a == b
+  (==) (Time       a) (Time       b) = a == b
+  (==) (Interval u v) (Interval x y) = (u == x) && (v == y)
+  (==) (Text       a) (Text       b) = a == b
+  (==) (Bool       a) (Bool       b) = a == b
+  (==) (Percent    a) (Percent    b) = a == b
+  (==) (Currency   a) (Currency   b) = a == b
+  (==) (NA) (NA) = true
+  (==) _ _ = false
+  (/=) a b = not (a == b)
+
+
+noFlags :: RegexFlags
+noFlags  = { global     : false
+           , ignoreCase : false
+           , multiline  : false
+           , sticky     : false
+           , unicode    : false }
+
+percentRegex :: String -> Boolean
+percentRegex = test $ regex """^-?\d+(\.\d+)?\%$""" noFlags
+
+currencyRegex :: String -> Boolean
+currencyRegex = test $ regex "^\\$?([0-9]{1,3},([0-9]{3},)*[0-9]{3}|[0-9]+)(.[0-9][0-9])?$" noFlags
+
+-- somewhat like ISO8601
+dateRegex :: String -> Boolean
+dateRegex = test $ regex """^(\d{4})-(\d{2})-(\d{2})T(\d{2})\:(\d{2})\:(\d{2})[+-](\d{2})\:(\d{2})$""" noFlags
+
+analyzeNum :: Number -> JSemantic
+analyzeNum n | floor n == n = Integral n
+analyzeNum n                = Fractional n
+
+foreign import parsePercent """function parsePercent (s) {
+  return s.replace("%", "") * 1}""" :: String -> Number
+
+foreign import parseCurrency """function parseCurrency (s) {
```

Comment:
Aside from `$`, you can catch all unicode currency symbols by looking for characters in the range 20A0–20CF.

Beyond that, this function seems fishy, because its type claims to be able to convert any string at all into a number (by the type I'd guess it computes the string length, which is the only way I can think of to reasonably satisfy that promise).